frontend/server/src/main/java/com/amazonaws/ml/mms/metrics/MetricCollector.java

```diff
@@ -30,71 +37,108 @@
-                                if (!error.isEmpty()) {
-                                    logger.error(error);
-                                }
-                            } catch (IOException e) {
-                                logger.error(null, e);
-                            }
-                        })
-                .start();
+            try (OutputStream os = p.getOutputStream()) {
+                IOUtils.write(getWorkerPids(workerMap), os, StandardCharsets.UTF_8);
+            }
 
-        return IOUtils.toString(p.getInputStream(), StandardCharsets.UTF_8);
-    }
+            new Thread(
+                            () -> {
+                                try {
+                                    String error =
+                                            IOUtils.toString(
+                                                    p.getErrorStream(), StandardCharsets.UTF_8);
+                                    if (!error.isEmpty()) {
+                                        logger.error(error);
+                                    }
+                                } catch (IOException e) {
+                                    logger.error("", e);
+                                }
+                            })
+                    .start();
 
-    @Override
-    public void run() {
-        try {
-            String metricJsonString = collect();
             MetricManager metricManager = MetricManager.getInstance();
-            metricManager.setMetrics(JsonUtils.GSON.fromJson(metricJsonString, LIST_TYPE));
-            loggerMetrics.info(metricManager.getMetrics());
+            try (BufferedReader reader =
+                    new BufferedReader(
+                            new InputStreamReader(p.getInputStream(), StandardCharsets.UTF_8))) {
+                // first line is system metrics
+                String line = reader.readLine();
+                if (line == null || line.isEmpty()) {
+                    logger.error("Expecting system metrics line, but received empty.");
+                    return;
+                }
+
+                List<Metric> metricsSystem = JsonUtils.GSON.fromJson(line, LIST_TYPE);
+                metricManager.setMetrics(metricsSystem);
+                loggerMetrics.info(metricsSystem);
+
+                // Collect process level metrics
+                while ((line = reader.readLine()) != null) {
+                    String[] tokens = line.split(":");
+                    if (tokens.length != 2) {
+                        continue;
+                    }
+
+                    Integer pid = Integer.valueOf(tokens[0]);
+                    WorkerThread worker = workerMap.get(pid);
+                    worker.setMemory(Long.parseLong(tokens[1]));
+                }
+            }
         } catch (IOException e) {
             logger.error("", e);
         }
     }
+
+    private String getWorkerPids(Map<Integer, WorkerThread> workerMap) {
```

Comment:
Passing OutputStream into this function can avoid create StringBuilder object:
    private void getWorkerPids(Map<Integer, WorkerThread> workerMap, OutputStream os) throws IOException {
        boolean first = true;
        for (Integer pid : workerMap.keySet()) {
            if (pid <= 0) {
                logger.warn("worker pid is not available yet.");
                continue;
            }
            if (first) {
                first = false;
            } else {
                IOUtils.write(",", os,  StandardCharsets.UTF_8);
            }
            IOUtils.write(pid.toString(), os,  StandardCharsets.UTF_8);
        }
    }